ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +175,424 @@ let order_f (_, vm_rec) =
+   with memory size: "vm_size" is planned on the "host":
+   1. For the "group", increase "vm_cnt" of the "host" by 1.
+   2. For each group, updates the host's size by substracting "vm_size". *)
+let update_spread_evenly_plan_pool_state vm_size group host pool_state =
+  VMGrpMap.mapi
+    (fun grp hosts_q ->
+      match grp = group with
+      | true ->
+          Q.adjust host
+            (fun (vm_cnt, h_size) -> (vm_cnt + 1, Int64.sub h_size vm_size))
+            hosts_q
+      | false ->
+          Q.adjust host
+            (fun (vm_cnt, h_size) -> (vm_cnt, Int64.sub h_size vm_size))
+            hosts_q
+    )
+    pool_state
+
+(* The no breach plan pool state determines the no breach evacuation planning
+   for anti-affinity VMs.
+   It's a VMGrpMap which contains "no breach plan state" for each VM anti-
+   affinity group.
+   "no breach plan state" has 2 elements:
+   1. a Psq Q which contains "no_resident" hosts for that group. (A "no
+      resident" host for a group is a host which has no running VMs from that
+      group resident on it.)
+   2. an int which is the number of "resident" hosts for each group. (A
+      "resident" host for a group is a host which has at least one running VM
+      from that group resident on it.)
+   Let's say the anti-affinity VM to be planned belongs to anti-affinity group
+   A. If for group A, the number of "resident" hosts is already 2 or greater
+   than 2, then we don't need to plan the VM on any host, if not, we will need
+   to check the host with the minimal free memory from the "no resident" hosts
+   queue, which is just the minimal host returned from "Q.min" on the "no
+   resident" hosts Psq of group A. *)
+let init_no_breach_plan_pool_state spread_evenly_plan_pool_state =
+  spread_evenly_plan_pool_state
+  |> VMGrpMap.map (fun hs ->
+         ( Q.filter (fun _h (vm_cnt, _h_size) -> vm_cnt = 0) hs
+         , hs |> Q.filter (fun _h (vm_cnt, _size) -> vm_cnt > 0) |> Q.size
+         )
+     )
+
+(* Update "no_breach_plan_pool_state" after a VM from anti-affinity "group"
+   with memory size: "vm_size" is planned on the "host":
+   1. For the "group", the "host" is removed from its "no_resident" hosts
+      queue, and increase its "resident_hosts_cnt" by 1.
+   2. For other groups, updates the host's size by substracting "vm_size" if
+      the host is in that group's "no_resident" hosts queue. *)
+let update_no_breach_plan_pool_state vm_size group host pool_state =
+  VMGrpMap.mapi
+    (fun grp (no_resident_hosts, resident_hosts_cnt) ->
+      match grp = group with
+      | true ->
+          (Q.remove host no_resident_hosts, succ resident_hosts_cnt)
+      | false ->
+          ( Q.update host
+              (Option.map (fun (vm_cnt, h_size) ->
+                   (vm_cnt, Int64.sub h_size vm_size)
+               )
+              )
+              no_resident_hosts
+          , resident_hosts_cnt
+          )
```

Comment:
| false ->
          let f = Option.map (fun (c, s) -> (c, (Int64.sub s vm_size))) in
          (Q.update host f no_resident_hosts, resident_hosts_cnt)